In [ ]:
from pathlib import Path
from google.colab import userdata

import subprocess
import pandas as pd

# token = userdata.get('GH_TOKEN')
# assert token, 'Add GH_TOKEN to Colab Secrets, then enable notebook access.'

token = "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"

repo = '/content/matmul_cuda'
auth_url = f'https://Sushil2006:{token}@github.com/Sushil2006/matmul_cuda.git'
clean_url = 'https://github.com/Sushil2006/matmul_cuda.git'

if Path(repo, '.git').is_dir():
    !git -C {repo} remote set-url origin {auth_url}
    !git -C {repo} pull --ff-only
else:
    !git clone {auth_url} {repo}

!git -C {repo} remote set-url origin {clean_url}
%cd /content/matmul_cuda/

Cloning into '/content/matmul_cuda'...
remote: Enumerating objects: 39, done.
remote: Counting objects: 100% (39/39), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 39 (delta 19), reused 33 (delta 13), pack-reused 0 (from 0)
Receiving objects: 100% (39/39), 327.71 KiB | 1.33 MiB/s, done.
Resolving deltas: 100% (19/19), done.
/content/matmul_cuda


In [7]:
import torch

p = torch.cuda.get_device_properties(0)

print("GPU:", torch.cuda.get_device_name(0))
print("Compute capability:", torch.cuda.get_device_capability(0))
print("VRAM (GiB):", round(p.total_memory / 2**30, 2))
print("SM count:", p.multi_processor_count)
print("Max threads / block:", p.max_threads_per_block)
print("Shared memory / block (KiB):", p.shared_memory_per_block // 1024)
print("Shared memory / SM (KiB):", p.shared_memory_per_multiprocessor // 1024)
print("Threads / block:", p.max_threads_per_block)
print("Threads / SM:", p.max_threads_per_multi_processor)

GPU: Tesla T4
Compute capability: (7, 5)
VRAM (GiB): 14.56
SM count: 40
Max threads / block: 1024
Shared memory / block (KiB): 48
Shared memory / SM (KiB): 64
Threads / block: 1024
Threads / SM: 1024


# Shared-Memory Tiled FP32 GEMM

## Experiment: tile-configuration sweep

- Goal: measure how `BM`, `BN`, and `BK` affect a shared-memory GEMM where one thread computes one output element.
- Matrix shape: `M=N=K=8192`; all values are FP32.
- Timing: CUDA events, one warm-up launch, then the average of three timed launches.
- Correctness: each result is checked against cuBLAS FP32 output.


In [ ]:
!nvcc -std=c++20 -O3 benchmark.cu kernels/naive.cu kernels/smem_tiled.cu kernels/block_tiled.cu -lcublas -o matmul

sizes = [8192]
configs = [
    (16, 16, 8),
    (16, 16, 16),
    (16, 16, 32),
    (32, 8, 16),
    (8, 32, 16),
    (32, 32, 8),
    (32, 32, 16),
    (32, 32, 32),
    (32, 32, 64),
    (32, 32, 128),
]
rows = []

for size in sizes:
    for bm, bn, bk in configs:
        print(f'running N={size} BM={bm} BN={bn} BK={bk}', flush=True)
        completed = subprocess.run(
            ['./matmul', '--kernel', 'smem', '--M', str(size), '--N', str(size), '--K', str(size),
             '--BM', str(bm), '--BN', str(bn), '--BK', str(bk)],
            capture_output=True, text=True, check=True)
        metrics = dict(item.split('=', 1) for item in completed.stdout.strip().splitlines()[-1].split())
        print(f'  time_ms={metrics["time_ms"]}', flush=True)
        rows.append({
            'N': size, 'BM': bm, 'BN': bn, 'BK': bk,
            'time_ms': float(metrics['time_ms']),
            'tflops': float(metrics['tflops']),
            'max_abs_error': float(metrics['max_abs_error']),
            'status': metrics['status'],
        })

df = pd.DataFrame(rows).sort_values(['N', 'time_ms']).reset_index(drop=True)
df

nvcc warning : Support for offline compilation for architectures prior to '<compute/sm/lto>_75' will be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
running N=8192 BM=16 BN=16 BK=8
  time_ms=2111.53
running N=8192 BM=16 BN=16 BK=16
  time_ms=1767.89
running N=8192 BM=16 BN=16 BK=32
  time_ms=1782.76
running N=8192 BM=32 BN=8 BK=16
  time_ms=1806.74
running N=8192 BM=8 BN=32 BK=16
  time_ms=2116.13
running N=8192 BM=32 BN=32 BK=8
  time_ms=2111.46
running N=8192 BM=32 BN=32 BK=16
  time_ms=1702.92
running N=8192 BM=32 BN=32 BK=32
  time_ms=1391.58
running N=8192 BM=32 BN=32 BK=64
  time_ms=1357.6
running N=8192 BM=32 BN=32 BK=128
  time_ms=1205.35


,N,BM,BN,BK,time_ms,tflops,max_abs_error,status
0,8192,32,32,128,1205.35,0.912189,0.0,PASS
1,8192,32,32,64,1357.60,0.809896,0.0,PASS
2,8192,32,32,32,1391.58,0.790116,0.0,PASS
3,8192,32,32,16,1702.92,0.645663,0.0,PASS
4,8192,16,16,16,1767.89,0.621933,0.0,PASS
5,8192,16,16,32,1782.76,0.616746,0.0,PASS
6,8192,32,8,16,1806.74,0.608562,0.0,PASS
7,8192,32,32,8,2111.46,0.520735,0.0,PASS
8,8192,16,16,8,2111.53,0.520719,0.0,PASS
9,8192,8,32,16,2116.13,0.519586,0.0,PASS


## Results and interpretation

- The best tested configuration was `(BM, BN, BK) = (32, 32, 128)`: `1205.35 ms` and `0.912 TFLOP/s`.
- For `32 x 32` tiles, increasing `BK` from `8` to `128` reduced runtime from `2111.46 ms` to `1205.35 ms` (`1.75x` faster).
- For fixed `BM` and `BN`, larger `BK` does not change global-memory reuse. It reduces tile-loop overhead and synchronization frequency.

```text
tile iterations = K / BK
barriers = 2 * K / BK

BK=8   -> 1024 tile iterations, 2048 barriers
BK=128 ->   64 tile iterations,  128 barriers
```

- Ignoring the final C store, input arithmetic intensity per tile is approximately:

```text
BM * BN / (2 * (BM + BN)) FLOP/byte

16 x 16 -> 4 FLOP/byte
32 x 32 -> 8 FLOP/byte
32 x 8  -> 3.2 FLOP/byte
```

- A `32 x 32` block has `1024` threads, equal to this T4's `1024 threads/SM` limit, so it uses one resident block per SM while filling the thread capacity.
- `(32, 32, 128)` uses `32 KiB` shared memory per block: `(32 + 32) * 128 * 4`; this fits the `48 KiB` per-block limit.


## Thread-tiled configuration sweep

- This kernel keeps shared-memory tiles but lets each thread compute a `TM x TN` output window in registers.
- The configs below vary `BK`, thread-tile shape, and output-tile shape without a full Cartesian-product sweep.
- Run the compile cell above after changing CUDA source files, then run this cell.


In [9]:
block_configs = [
    (64, 64, 8, 4, 4),
    (64, 64, 16, 4, 4),
    (64, 64, 32, 4, 4),
    (64, 64, 8, 8, 4),
    (64, 64, 8, 4, 8),
    (128, 64, 8, 8, 4),
    (64, 128, 8, 4, 8),
    (64, 64, 8, 1, 16),
    (64, 64, 8, 16, 1),
    (64, 64, 64, 4, 4),
    (128, 128, 32, 8, 8),
]
block_rows = []

for size in sizes:
    for bm, bn, bk, tm, tn in block_configs:
        print(f'running N={size} BM={bm} BN={bn} BK={bk} TM={tm} TN={tn}', flush=True)
        completed = subprocess.run(
            ['./matmul', '--kernel', 'block', '--M', str(size), '--N', str(size), '--K', str(size),
             '--BM', str(bm), '--BN', str(bn), '--BK', str(bk), '--TM', str(tm), '--TN', str(tn)],
            capture_output=True, text=True, check=True)
        metrics = dict(item.split('=', 1) for item in completed.stdout.strip().splitlines()[-1].split())
        print(f'  time_ms={metrics["time_ms"]}', flush=True)
        block_rows.append({
            'N': size, 'BM': bm, 'BN': bn, 'BK': bk, 'TM': tm, 'TN': tn,
            'time_ms': float(metrics['time_ms']),
            'tflops': float(metrics['tflops']),
            'max_abs_error': float(metrics['max_abs_error']),
            'status': metrics['status'],
        })

block_df = pd.DataFrame(block_rows).sort_values(['N', 'time_ms']).reset_index(drop=True)
block_df

running N=8192 BM=64 BN=64 BK=8 TM=4 TN=4


CalledProcessError: Command '['./matmul', '--kernel', 'block', '--M', '8192', '--N', '8192', '--K', '8192', '--BM', '64', '--BN', '64', '--BK', '8', '--TM', '4', '--TN', '4']' returned non-zero exit status 1.